src/coreclr/jit/llvm.cpp

```diff
@@ -2713,6 +2754,34 @@ void Llvm::lowerToShadowStack()
     }
 }
 
+void Llvm::LowerPromotedFields()
+{
+    for (BasicBlock* _currentBlock : _compiler->Blocks())
+    {
+        _currentRange = &LIR::AsRange(_currentBlock);
+        for (GenTree* node : CurrentRange())
+        {
+            if (node->OperIs(GT_LCL_VAR))
+            {
+                LclVarDsc* varDsc = _compiler->lvaGetDesc(node->AsLclVarCommon());
+                if (isIndependentPromotedStrucField(varDsc))
```

Comment:
It's the other way around: dependently promoted locals need special treatment. Independently promoted ones don't, generally speaking (there is one special case that needs to be taken of for them).

Reply:
I was following https://github.com/dotnet/runtimelab/issues/1859#issuecomment-1060065739 but maybe I didn't understand that properly, or thinking has changed.

Reply:
I don't think it has...?

> From the backend perspective, [we should treat independently promoted fields as ordinary locals], and [dependently promoted ones as being modified with stores to their fields (the field LCL_VARs become "effective" LCL_FLDs)]. I think we should undo dependent promotion in lowering and replace all of the fields with actual LCL_FLDs, for simplicity.

Reply:
It was the argument bit I was struggling with, "The backend must also ensure that fields from arguments get "homed" into their locations properly. For independent promotion, I propose we lower them into the following:" and  I think the Equals method on `DateTime`s .  Let me get the IR.

Reply:
If we have 
```
Trees after Build SSA representation

-----------------------------------------------------------------------------------------------------------------------------------------
BBnum BBid ref try hnd preds           weight    lp [IL range]     [jump]      [EH region]         [flags]
-----------------------------------------------------------------------------------------------------------------------------------------
BB01 [0000]  1                             1       [000..01F)        (return)                     i LIR
-----------------------------------------------------------------------------------------------------------------------------------------

------------ BB01 [000..01F) (return), preds={} succs={}
N001 (  3,  2) [000035] ------------        t35 =    LCL_VAR   long   V05 tmp2         u:1 (last use) $80
N002 (  1,  4) [000036] ------------        t36 =    CNS_LNG   long   0x3fffffffffffffff $100
                                                  /--*  t35    long
                                                  +--*  t36    long
N003 (  8, 10) [000037] ------------        t37 = *  AND       long   $140
                                                  /--*  t37    long
N005 (  8, 10) [000025] DA----------              *  STORE_LCL_VAR long   V04 tmp1         d:1
N001 (  1,  1) [000026] ------------        t26 =    LCL_VAR   long   V04 tmp1         u:1 (last use) $140
N002 (  3,  2) [000041] ------------        t41 =    LCL_VAR   long   V06 tmp3         u:1 (last use) $81
N003 (  1,  4) [000042] ------------        t42 =    CNS_LNG   long   0x3fffffffffffffff $100
                                                  /--*  t41    long
                                                  +--*  t42    long
N004 (  8, 10) [000043] ------------        t43 = *  AND       long   $141
                                                  /--*  t26    long
                                                  +--*  t43    long
N005 ( 16, 15) [000028] ------------        t28 = *  EQ        int    $180
                                                  /--*  t28    int
N006 ( 17, 16) [000016] ------------              *  RETURN    int    $1c0
```
Then `V5` (`0x000002cc7375dfe8 [V5: TYP_LONG-"field V01._dateData (fldOffset=0x0)"]`) is not marked as a `isLlvmFrameLocal` has no `lvLlvmArgNum` and we fail the function compilation for this method

https://github.com/dotnet/runtimelab/blob/6629e07d7244c49f16903b335839adcebe17b2a2/src/coreclr/jit/llvm.cpp#L1862

Reply:
Yeah, so, the idea with parameters was to do this:
```c
for (unsigned lclNum = 0; lclNum < lvaCount; lclNum++)
{
    LclVarDsc* varDsc = lvaGetDesc(lclNum);
    if (varDsc->lvParam && (lvaGetPromotionType(varDsc) == PROMOTION_TYPE_INDEPENDENT))
    {
        for (unsigned index = 0; index < varDsc->lvFieldCnt; index++)
        {
            unsigned   fieldLclNum = varDsc->lvFieldStart + index;
            LclVarDsc* fieldVarDsc = lvaGetDesc(fieldLclNum);
            if (fieldVarDsc->lvRefCnt(RCS_NORMAL) != 0)
            {
                fgEnsureFirstBBIsScratch();
                LIR::Range& firstBlockRange = LIR::AsRange(fgFirstBB);
                
                GenTree* fieldValue = gtNewLclFldNode(lclNum, fieldVarDsc->TypeGet(), fieldVarDsc->lvFieldOffset);
                GenTree* fieldStore = gtNewStoreLclVar(fieldLclNum, fieldValue);
                firstBlockRange.InsertAtBeginning(fieldStore);
                firstBlockRange.InsertAtBeginning(fieldValue);
            }
            
            varDsc->lvParam = 0;
        }
    }
}
```
That way the promoted fields should get initialized appropriately.

Reply:
We end up with 
```
------------ BB01 [000..01F) (return), preds={} succs={}
               [000050] ------------        t50 =    LCL_FLD   long   V02 arg2         [+0]
                                                  *    long   V02._dateData (offs=0x00) -> V06 tmp3
                                                  /--*  t50    long
               [000051] DA----------              *  STORE_LCL_VAR long   V06 tmp3         d:2
```
but for `LCL_FLD` I have a couple of questions, for `V02` are we expecting it to pass this assert:
```
assert(isLlvmFrameLocal(varDsc));
```
?  Currently as there are no references it fails.  Secondly for `LCL_FLD` to be compiled, it needs to load from an address in out `m_allocas` but that will be empty.

Reply:
>  Currently as there are no references it fails

No references as in the refcount on the parent is zero? Doesn't seem right, the recounter should see the parent...

That said, some metadata in my sample also needs to be cleared for the local table to say consistent. Also, there is a bug with the `lvParam` setting (cleared on the parent instead of the field):
```diff
LclVarDsc* varDsc = lvaGetDesc(lclNum);
if (varDsc->lvParam && (lvaGetPromotionType(varDsc) == PROMOTION_TYPE_INDEPENDENT))
{
    for (unsigned index = 0; index < varDsc->lvFieldCnt; index++)
    {
        unsigned   fieldLclNum = varDsc->lvFieldStart + index;
        LclVarDsc* fieldVarDsc = lvaGetDesc(fieldLclNum);
        if (fieldVarDsc->lvRefCnt(RCS_NORMAL) != 0)
        {
            fgEnsureFirstBBIsScratch();
            LIR::Range& firstBlockRange = LIR::AsRange(fgFirstBB);
            
            GenTree* fieldValue = gtNewLclFldNode(lclNum, fieldVarDsc->TypeGet(), fieldVarDsc->lvFieldOffset);
            GenTree* fieldStore = gtNewStoreLclVar(fieldLclNum, fieldValue);
            firstBlockRange.InsertAtBeginning(fieldStore);
            firstBlockRange.InsertAtBeginning(fieldValue);
        }
        
+       fieldVarDsc->lvIsStructField = false;
+       fieldVarDsc->lvParentLcl     = BAD_VAR_NUM;
+       fieldVarDsc->lvParam         = false;
    }

+   varDsc->lvPromoted   = false;
+   varDsc->lvFieldStart = BAD_VAR_NUM;
+   varDsc->lvFieldCnt   = 0;
}
```

Reply:
Awesome, I don't know which one of those did it, but all good now, at least for this method.  Thank you.
```ll
define i8 @"S_P_CoreLib_System_Collections_Generic_GenericEqualityComparer_1<S_P_CoreLib_System_DateTime>__Equals"(i8* %0, %"[S.P.CoreLib]System.DateTime" %1, %"[S.P.CoreLib]System.DateTime" %2) {
Prolog:
  %3 = alloca %"[S.P.CoreLib]System.DateTime", align 8
  store %"[S.P.CoreLib]System.DateTime" %1, %"[S.P.CoreLib]System.DateTime"* %3, align 1
  %4 = alloca %"[S.P.CoreLib]System.DateTime", align 8
  store %"[S.P.CoreLib]System.DateTime" %2, %"[S.P.CoreLib]System.DateTime"* %4, align 1
  br label %5

5:                                                ; preds = %Prolog
  %6 = bitcast %"[S.P.CoreLib]System.DateTime"* %4 to i8*
  %7 = getelementptr i8, i8* %6, i16 0
  %8 = bitcast i8* %7 to i64*
  %9 = load i64, i64* %8, align 8
  %10 = bitcast %"[S.P.CoreLib]System.DateTime"* %3 to i8*
  %11 = getelementptr i8, i8* %10, i16 0
  %12 = bitcast i8* %11 to i64*
  %13 = load i64, i64* %12, align 8
  %and = and i64 %13, 4611686018427387903
  %and1 = and i64 %9, 4611686018427387903
  %14 = icmp eq i64 %and, %and1
  %15 = zext i1 %14 to i8
  ret i8 %15
}
```

Reply:
Nice!